### PR TITLE
Modernize scaffold generator

### PR DESCRIPTION
**Pull request by dhh**
The default scaffold generator should be encouraging the use of partials. It also shouldn't try so hard to style things. We can do such styling as a build-on-top via CSS framework gems, like tailwindcss-rails.

Co-authored-by: Abhay Nikam <nikam.abhay1@gmail.com>

### CODE REVIEW

**Comment by dhh** [comment]
If anyone would like to help me wrap this up with correcting tests, I'd appreciate it!

**Comment by Diogomartf** [comment]
> If anyone would like to help me wrap this up with correcting tests, I'd appreciate it!

I would like to help, but how do I know which tests I need to correct? All that are failing?  
There are some that don't seem related.

Sorry if that might be obvious but it's my first time here ✌️

**Comment by dhh** [comment]
Thanks @Diogomartf! Yes, basically get this PR passing. There are some failing tests in the general suite that won't stop the overall build from passing. You can ignore that. Just all the ones focused on these changes ✌️❤️

File: railties/lib/rails/generators/erb/scaffold/templates/new.html.erb.tt [hunk 1]
```diff
@@ -1,5 +1,9 @@
-<h1>New <%= singular_table_name.titleize %></h1>
+<h1>New <%= singular_table_name %></h1>
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/erb/scaffold/templates/new.html.erb.tt:1]: Why was titleize removed from here? Do we want to generate `New user_account`?

File: railties/lib/rails/generators/rails/scaffold_controller/templates/api_controller.rb.tt [hunk 2]
```diff
@@ -4,7 +4,7 @@ require_dependency "<%= namespaced_path %>/application_controller"
 <% end -%>
 <% module_namespacing do -%>
 class <%= controller_class_name %>Controller < ApplicationController
-  before_action :set_<%= singular_table_name %>, only: [:show, :update, :destroy]
+  before_action :set_<%= singular_table_name %>, only: %w[ show update destroy ]
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/rails/scaffold_controller/templates/api_controller.rb.tt:7]: should not it be `%i`?

File: railties/lib/rails/generators/erb/scaffold/templates/new.html.erb.tt (hunk 1 above)
Comment (dhh) [reply railties/lib/rails/generators/erb/scaffold/templates/new.html.erb.tt:1]: Ahh, I was just looking to get rid of the capitalization. I guess we'd need to do #human_name instead?

File: railties/lib/rails/generators/erb/scaffold/templates/show.html.erb.tt [hunk 3]
```diff
@@ -1,19 +1,10 @@
...
+<%%= render @<%= singular_table_name %> %>
 
-<% end -%>
-<%%= link_to 'Edit', edit_<%= singular_table_name %>_path(@<%= singular_table_name %>) %> |
-<%%= link_to 'Back', <%= index_helper %>_path %>
+<div>
+  <%%= link_to "Edit this <%= human_name.downcase %>", edit_<%= singular_table_name %>_path(@<%= singular_table_name %>) %> |
+  <%%= link_to "Back to <%= human_name.pluralize.downcase %>", <%= index_helper %>_path %>
+
+  <%%= button_to "Destroy this <%= human_name.downcase %>", <%= singular_table_name %>_path(@<%= singular_table_name %>), method: :delete %>
```
Comment (abhaynikam) [inline railties/lib/rails/generators/erb/scaffold/templates/show.html.erb.tt:9]: Maybe `link_to` would be better? 

```suggestion
  <%%=link_to "Destroy this <%= human_name.downcase %>", <%= singular_table_name %>_path(@<%= singular_table_name %>), method: :delete, data: { confirm: 'Are you sure?' } %>
```

Comment (dhh) [reply railties/lib/rails/generators/erb/scaffold/templates/show.html.erb.tt:9]: We are looking to phase out links that perform DELETE. It's not a good, safe pattern.

File: railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt [hunk 4]
```diff
@@ -2,19 +2,18 @@
...
       <%- unless options[:skip_turbolinks] -%>
-    <%%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
-    <%%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <%%= stylesheet_link_tag "application", "data-turbolinks-track": "reload" %>
+    <%%= javascript_pack_tag "application", "data-turbolinks-track": "reload" %>
       <%- else -%>
-    <%%= stylesheet_link_tag 'application', media: 'all' %>
-    <%%= javascript_pack_tag 'application' %>
+    <%%= stylesheet_link_tag "application" %>
+    <%%= javascript_pack_tag "application", media: "all" %>
```
Comment (hahmed) [inline railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:16]:
```suggestion
    <%%= javascript_pack_tag "application" %>
```

Looks like we can drop the media: "all" as it's not needed for javascript tags

File: railties/test/generators/scaffold_controller_generator_test.rb [hunk 5]
```diff
@@ -221,14 +221,11 @@ def test_model_name_option
     end
 
     assert_file "app/views/admin/users/index.html.erb" do |content|
-      assert_match("'Show', [:admin, user]", content)
-      assert_match("'Edit', edit_admin_user_path(user)", content)
-      assert_match("'Destroy', [:admin, user]", content)
-      assert_match("'New User', new_admin_user_path", content)
+      assert_match("'New user', new_admin_user_path", content)
```
Comment (hahmed) [inline railties/test/generators/scaffold_controller_generator_test.rb:224]:
```suggestion
      assert_match("\"New user\", new_admin_user_path", content)
```

(double quote fixes the test)

File: railties/test/generators/scaffold_controller_generator_test.rb [hunk 6]
```diff
@@ -221,14 +221,11 @@ def test_model_name_option
...
     end
 
     assert_file "app/views/admin/users/new.html.erb" do |content|
-      assert_match("'Back', admin_users_path", content)
+      assert_match("\"Back to user\", admin_users_path", content)
```
Comment (hahmed) [inline railties/test/generators/scaffold_controller_generator_test.rb:228]:
```suggestion
      assert_match("\"Back to users\", admin_users_path", content)
```

the s was missing, in users which fixes this test.

File: railties/test/generators/scaffold_controller_generator_test.rb (hunk 5 above)
Comment (hahmed) [reply railties/test/generators/scaffold_controller_generator_test.rb:224]: test failing:


ScaffoldControllerGeneratorTest#test_model_name_option [test/generators/scaffold_controller_generator_test.rb:223]:
--
  | Expected /'New\ user',\ new_admin_user_path/ to match "<p id=\"notice\"><%= notice %></p>\n\n<h1>User</h1>\n\n<div id=\"users\">\n  <%= render @users %>\n</div>\n\n<%= link_to \"New user\", new_admin_user_path %>\n".

File: railties/test/generators/scaffold_generator_test.rb [hunk 7]
```diff
@@ -282,10 +282,7 @@ def test_scaffold_with_namespace_on_invoke
 
     # Views
     assert_file "app/views/admin/roles/index.html.erb" do |content|
-      assert_match("'Show', admin_role", content)
-      assert_match("'Edit', edit_admin_role_path(admin_role)", content)
-      assert_match("'Destroy', admin_role", content)
-      assert_match("'New Admin Role', new_admin_role_path", content)
+      assert_match("'New role', new_admin_role_path", content)
```
Comment (jonathanhefner) [inline railties/test/generators/scaffold_generator_test.rb:285]:
```suggestion
      assert_match(%("New role", new_admin_role_path), content)
```

**Comment by jonathanhefner** [comment]
What do you think about extracting the flash message to a partial as well?  That would make it easier to style with utility classes a la Tailwind.  We could implement `FlashHash#to_partial_path` in a follow-up PR so that the render call could be simply `<%= render flash %>`.

**Comment by dhh** [comment]
I think that's going too far. The reason we didn't use partials the first time around is that the purpose with these scaffolds is soft on-ramp learning. Flash doesn't warrant an extra file for that.

File: activerecord/CHANGELOG.md [hunk 8]
```diff
@@ -38,7 +38,6 @@
 
     ```ruby
     class User < ApplicationRecord
-      has_many :bookmarks
```
Comment (abhaynikam) [inline activerecord/CHANGELOG.md:41]: The association was used below to demonstrate usage of strict_loading on a Model. Is the removal intentional or removed by mistake while rebasing the PR?

Comment (dhh) [reply activerecord/CHANGELOG.md:41]: mistake, adding back

Comment (abhaynikam) [reply activerecord/CHANGELOG.md:41]: Raised a PR for it: https://github.com/rails/rails/pull/41329. Please feel free to close it if you have added the change back 😊

**Comment by yshmarov** [comment]
can't wait to use the new scaffold!

**Pull request merged**